ragnarok_formats/src/model.rs

```diff
@@ -57,27 +57,84 @@ where
 
 #[derive(Debug, ByteConvertable)]
 #[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
-pub struct PositionKeyframeData {
+pub struct ScaleKeyframeData {
     pub frame: u32,
-    pub position: Point3<f32>,
+    pub scale: Vector3<f32>,
+    reserved: f32,
 }
-
 #[derive(Clone, Debug, ByteConvertable)]
 #[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
 pub struct RotationKeyframeData {
     pub frame: u32,
     pub quaternions: Quaternion<f32>,
 }
 
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TranslationKeyframeData {
+    pub frame: u32,
+    pub translation: Vector3<f32>,
+    reserved: f32,
+}
+
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TexturesKeyframeData {
+    pub texture_index: u32,
+    #[new_derive]
+    pub texture_keyframe_count: u32,
+    #[repeating(texture_keyframe_count)]
+    pub texture_keyframes: Vec<TextureKeyframeData>,
+}
+
+/// List of texture operation types.
+/// See: https://rathena.org/board/topic/127587-rsm2-file-format/
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub enum TextureOperation {
+    /// Texture translation on the X axis. The texture is tiled.
+    TranslationX,
+    /// Texture translation on the Y axis. The texture is tiled.
+    TranslationY,
+    /// Texture multiplication on the X axis. The texture is tiled.
+    ScaleX,
+    /// Texture multiplication on the Y axis. The texture is tiled
+    ScaleY,
+    /// Texture rotation around (0, 0). The texture is not tiled.
+    Rotation,
+}
+
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TextureKeyframeData {
+    pub operation_type: TextureOperation,
+    #[new_derive]
+    pub frame_count: u32,
+    #[repeating(frame_count)]
+    pub texture_frames: Vec<TextureFrameData>,
+}
+
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TextureFrameData {
+    pub frame: u32,
+    pub operation_value: f32,
+}
+
 #[derive(Debug, ByteConvertable)]
 #[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
 pub struct FaceData {
+    #[version_equals_or_above(2, 2)]
+    pub length: Option<u32>,
     pub vertex_position_indices: [u16; 3],
     pub texture_coordinate_indices: [u16; 3],
     pub texture_index: u16,
     pub padding: u16,
     pub two_sided: i32,
     pub smooth_group: i32,
+    #[version_equals_or_above(2, 2)]
+    #[repeating_expr(0.max((length.unwrap() as usize - 24) / 4))]
```

Comment:
My bad, I just realized the `0.max` is also redundant. The type is already `usize` so it can't be negative